### PR TITLE
On predicate

### PR DESCRIPTION
**Pull request by gsakkis**
[Allow on to be a predicate](https://github.com/hynek/stamina/discussions/63)

### CODE REVIEW

**Comment by hynek** [comment]
@dhirschfeld @estahn Before I dive into this[^1] – opinions?

[^1]: I really need to get out _attrs_ 24.1 before starting something new 🫠

File: src/stamina/_core.py [hunk 1]
```diff
@@ -384,12 +393,20 @@ def from_params(
...
         kw: dict[str, object],
     ) -> _RetryContextIterator:
+        if (
+            isinstance(on, type)
+            and issubclass(on, BaseException)
+            or isinstance(on, tuple)
+        ):
+            _retry = _t.retry_if_exception_type(on)
+        else:
+            _retry = _t.retry_if_exception(on)
```
Comment (dhirschfeld) [inline src/stamina/_core.py:403]: Doing 3 type checks seems more complex than the example in the issue?
```suggestion
        if isinstance(on, Callable):
            _retry = _t.retry_if_exception(on)
        else:
            _retry = _t.retry_if_exception(lambda exc: isinstance(exc, on))
```

Comment (gsakkis) [reply src/stamina/_core.py:403]: This doesn't work (correctly), classes are callables too.

File: src/stamina/_core.py [hunk 2]
```diff
@@ -51,10 +60,14 @@ async def _smart_sleep(delay: float) -> None:
 
 T = TypeVar("T")
 P = ParamSpec("P")
+# for backwards compatibility with Python<3.10
+On = Union[
+    Type[Exception], Tuple[Type[Exception], ...], Callable[[Exception], bool]
+]
```
Comment (dhirschfeld) [inline src/stamina/_core.py:66]: I'm not a big fan of the name. `on: On` doesn't convey any meaning to the reader of the code.

My own inclination would be to try and give some meaning to the name - e.g. `on: exc_or_predicate`. It's not perfect, but I think it's probably better than conveying no meaning in the name.

**Review by dhirschfeld — APPROVED** [review]
Just some stylistic comments which @hynek can decide upon, otherwise I think it looks good.

File: src/stamina/_core.py (hunk 1 above)
Comment (dhirschfeld) [reply src/stamina/_core.py:403]: Hmm, true 🤔
I guess that's about the cleanest way then...

**Comment by dhirschfeld** [comment]
Resolves #63

File: src/stamina/_core.py (hunk 2 above)
Comment (gsakkis) [reply src/stamina/_core.py]: Renamed to `ExcOrPredicate`

**Comment by estahn** [comment]
Does this need documentation or an example for the docs?

**Comment by hynek** [comment]
> Does this need documentation or an example for the docs?

impeccable timing :D

**Comment by hynek** [comment]
I've just added the docs and am heading to the gym. If y'all could have look at them until I'm back that would be appreciated.

**Comment by estahn** [comment]
> I've just added the docs and am heading to the gym. If y'all could have look at them until I'm back that would be appreciated: https://py-stamina--70.org.readthedocs.build/en/70/

Looks good to me, although I realise it doesn't fully support my use-case I think.

We integrate with a large number of different API vendors and each have their own quirks. One of them being the handling of quota's. Some of those API's will respond with a 429 and provide quota details like time to next retry. I figured I could use this information as the basis for the retry mechanics, but `wait` is not a callable.

Similar to this example.
https://www.seelk.co/blog/efficient-client-side-handling-of-api-throttling-in-python-with-tenacity/

As for this PR 👍🏼 🚀

**Comment by hynek** [comment]
> I realise it doesn't fully support my use-case I think.
> 
> We integrate with a large number of different API vendors and each have their own quirks. One of them being the handling of quota's. Some of those API's will respond with a 429 and provide quota details like time to next retry. I figured I could use this information as the basis for the retry mechanics, but `wait` is not a callable.

So what you want is being about to return a float that means "yes, but at least _this_ long"?

Feel free to open a discussion about it, ideally with a simplified requirement vs whole blog post. 😇

> As for this PR 👍🏼 🚀

Thanks!

**Comment by estahn** [comment]
@hynek Thanks for merging. Is there anything else required before releasing this?

**Pull request merged**